Alternative lisp modules

Uthar opened the pull request:
###### Description of changes

Alternative lisp-modules implementation. 

See discussion https://github.com/NixOS/nixpkgs/issues/155851

Summary of changes:

- Fixed loading of FASLs from the Nix store, which fixes #142604
- Faster way of importing Quicklisp releases - currently imports all systems directly from QL's `systems.txt`
- Added support for specifying Java libraries (ABCL)
- Added the `{abcl,ecl,ccl,sbcl}WithPackages` functions, which produce self-contained store closures, compared to the `common-lisp.sh` solution which is based on shell environments.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

7c6f434c commented on pkgs/development/lisp-modules/default.nix:
```diff
@@ -0,0 +1,3 @@
+{ pkgs ? import <nixpkgs> { }, ... }:
+
+pkgs.callPackage ./nix-cl.nix { }
```
What this level of indirection is used for?

Uthar replied:
it's unneeded, removed it now in https://github.com/NixOS/nixpkgs/commit/690b62640e38bbc1e9310b3532741fbd9c6cc58c

7c6f434c commented:
(I guess if this renamed the current `lispPackages` to `lispPackagesLegacy` and added the new version side by side as new `lispPackages`, it would be a more readable diff in GitHub… then we can dropping the legacy after a few weeks of trying things out can be discussed separately, and will have a pure-removal easy to review diff)

Uthar commented:
> (I guess if this renamed the current `lispPackages` to `lispPackagesLegacy` and added the new version side by side as new `lispPackages`, it would be a more readable diff in GitHub… then we can dropping the legacy after a few weeks of trying things out can be discussed separately, and will have a pure-removal easy to review diff)

Thank you for the suggestion, I will fix this later this evening. In the meantime the second commit itself contains only additions, https://github.com/NixOS/nixpkgs/pull/170149/commits/e18582b985ceb0c509874941c969b97efbe1233d, so you could look there if you want.

Uthar committed:
add alternative lisp-modules implementation

based on https://github.com/Uthar/nix-cl/commit/4c4cef812e67677a55a2a949e49dd109ae1fcf93

Uthar commented:
I added it as `pkgs.lispPackages_new` without touching the current code, this might be cleaner, do you agree?

Uthar commented:
Is `builtins.fetchTarball` supposed to be avoided? It seems to force some eager evaluation of the `src` attribute. The CI was complaining about it. `pkgs.fetchurl` doesn't have this problem.

Uthar commented:
- Removed the asd conflict resolution code because it was too slow.
- Fixed eval time of big packages by removing slow code, so could remove the generic-cl hax from sqlite.lisp

Uthar commented:
added a test to locally check if packages build

Uthar commented:
switched to fetchzip to prevent downloading during evaluation

Uthar commented:
pulled out the working makeWrapper from an older nixpkgs commit

7c6f434c commented on pkgs/development/lisp-modules-new/doc/nix-cl.md:
```diff
@@ -0,0 +1,98 @@
+
+## Pinning down the exact commits of libraries
+
+Sometimes, a bug is fixed upstream but is not yet available in package
+repositories such as Quicklisp or Ultralisp. The users have to wait
```
I would probably say «a fix is available» if you want to mention Ultralisp — «fixed upstream» sounds like putting the fix into the main branch of the repo, and Ultralisp usually picks up things literally in a few minutes…

7c6f434c commented:
It's a pure addition, and I agree it is cleaner than the existing code, so let's merge it and see how much adapting the existing workflows need and based on that make a breaking-replacement warning/schedule later

Uthar commented:
Awesome, yes let's see if we like it.
